pkg/apis/configuration/v1/kongingress_types.go

```diff
@@ -48,6 +48,64 @@ type KongIngressList struct {
+	//+kubebuilder:validation:Minimum=0
+	WriteTimeout *int `json:"write_timeout,omitempty" yaml:"write_timeout,omitempty"`
+}
+
+// KongIngressRoute contains KongIngress route configuration
+//+ It contains the subset of go-kong.kong.Route fields supported by kongstate.Route.overrideByKongIngress
+type KongIngressRoute struct {
+	Methods                 []*string           `json:"methods,omitempty" yaml:"methods,omitempty"`
+	Headers                 map[string][]string `json:"headers,omitempty" yaml:"headers,omitempty"`
+	Protocols               []*KongProtocol     `json:"protocols,omitempty" yaml:"protocols,omitempty"`
+	RegexPriority           *int                `json:"regex_priority,omitempty" yaml:"regex_priority,omitempty"`
+	StripPath               *bool               `json:"strip_path,omitempty" yaml:"strip_path,omitempty"`
+	PreserveHost            *bool               `json:"preserve_host,omitempty" yaml:"preserve_host,omitempty"`
+	HTTPSRedirectStatusCode *int                `json:"https_redirect_status_code,omitempty" yaml:"https_redirect_status_code,omitempty"`
+	//+kubebuilder:validation:Enum=v0;v1
+	PathHandling      *string   `json:"path_handling,omitempty" yaml:"path_handling,omitempty"`
+	SNIs              []*string `json:"snis,omitempty" yaml:"snis,omitempty"`
+	RequestBuffering  *bool     `json:"request_buffering,omitempty" yaml:"request_buffering,omitempty"`
+	ResponseBuffering *bool     `json:"response_buffering,omitempty" yaml:"response_buffering,omitempty"`
+}
+
+// KongIngressUpstream contains KongIngress upstream configuration
+//+ It contains the subset of go-kong.kong.Upstream fields supported by kongstate.Upstream.overrideByKongIngress
+type KongIngressUpstream struct {
+	HostHeader *string `json:"host_header,omitempty" yaml:"host_header,omitempty"`
+	//+kubebuilder:validation:Enum=round-robin;consistent-hashing;least-connections
+	Algorithm *string `json:"algorithm,omitempty" yaml:"algorithm,omitempty"`
+	//+kubebuilder:validation:Minimum=10
+	Slots              *int              `json:"slots,omitempty" yaml:"slots,omitempty"`
+	Healthchecks       *kong.Healthcheck `json:"healthchecks,omitempty" yaml:"healthchecks,omitempty"`
+	HashOn             *string           `json:"hash_on,omitempty" yaml:"hash_on,omitempty"`
+	HashFallback       *string           `json:"hash_fallback,omitempty" yaml:"hash_fallback,omitempty"`
+	HashOnHeader       *string           `json:"hash_on_header,omitempty" yaml:"hash_on_header,omitempty"`
+	HashFallbackHeader *string           `json:"hash_fallback_header,omitempty" yaml:"hash_fallback_header,omitempty"`
+	HashOnCookie       *string           `json:"hash_on_cookie,omitempty" yaml:"hash_on_cookie,omitempty"`
+	HashOnCookiePath   *string           `json:"hash_on_cookie_path,omitempty" yaml:"hash_on_cookie_path,omitempty"`
+	// TODO status of this in existing code is unclear. While we supported a raw dump from KongIngress.upstream
```

Comment:
This seems to relate to the above TODO, do we have a tracking issue for this?

Reply:
d543d679